src/pdp/validation.ts

```diff
@@ -191,3 +192,104 @@ export function validateFindPieceResponse (value: unknown): FindPieceResponse {
     piece_cid: obj.piece_cid // Keep legacy field if it exists
   }
 }
+
+/**
+ * Converts and validates individual proof set root data
+ * Returns null if validation fails
+ *
+ * @param value - The value to validate and convert
+ * @returns Converted ProofSetRootData or null if invalid
+ */
+export function asProofSetRootData (value: unknown): ProofSetRootData | null {
+  if (typeof value !== 'object' || value == null) {
+    return null
+  }
+
+  const obj = value as Record<string, unknown>
+
+  // Required fields
+  if (typeof obj.rootId !== 'number') {
+    return null
+  }
+  if (typeof obj.rootCid !== 'string') {
+    return null
+  }
+  if (typeof obj.subrootCid !== 'string') {
+    return null
+  }
+  if (typeof obj.subrootOffset !== 'number') {
+    return null
+  }
+
+  // Convert CIDs to CommP objects
+  const rootCid = asCommP(obj.rootCid)
+  const subrootCid = asCommP(obj.subrootCid)
+  if (rootCid == null || subrootCid == null) {
+    return null
+  }
+
+  return {
+    rootId: obj.rootId,
+    rootCid,
+    subrootCid,
+    subrootOffset: obj.subrootOffset
+  }
+}
+
+/**
+ * Converts and validates proof set data
+ * Returns null if validation fails
+ *
+ * @param value - The value to validate and convert
+ * @returns Converted ProofSetData or null if invalid
+ */
+export function asProofSetData (value: unknown): ProofSetData | null {
+  if (typeof value !== 'object' || value == null) {
+    return null
+  }
+
+  const obj = value as Record<string, unknown>
+
+  // Required field - id
+  if (typeof obj.id !== 'number') {
+    return null
+  }
+
+  // Required field - roots (array of ProofSetRootData)
+  if (!Array.isArray(obj.roots)) {
+    return null
+  }
+
+  const convertedRoots: ProofSetRootData[] = []
+  for (const root of obj.roots) {
+    const convertedRoot = asProofSetRootData(root)
+    if (convertedRoot == null) {
+      return null
+    }
+    convertedRoots.push(convertedRoot)
+  }
+
+  // Required field - nextChallengeEpoch
+  if (typeof obj.nextChallengeEpoch !== 'number') {
+    return null
+  }
+
+  return {
+    id: obj.id,
+    roots: convertedRoots,
+    nextChallengeEpoch: obj.nextChallengeEpoch
+  }
+}
+
+/**
+ * Validates and returns a ProofSetData
+ * @param value - The value to validate and convert
+ * @throws Error if validation fails
+ */
+export function validateProofSetData (value: unknown): ProofSetData {
```

Comment:
ditch this function and just do what you're doing inside this where you call it, so we only have an `as`, not a `validate` for the cases where we always want to transform